### PR TITLE
Outline project 3

### LINKED ISSUE #37: Write outline for project 3
Just getting the outline done helps write the code.

### PR DESCRIPTION
**Pull request by brson**
This builds on https://github.com/pingcap/talent-plan/pull/36.

This project has been in flux, I made a bunch of calls here on what it is about, and I think it has become pretty substantial. Anybody who completes this course is going to be on their way to writing a real database.

Closes #37.

### CODE REVIEW

**Comment by brson** [comment]
@sticnarf I've pushed a few sections of text for this project. I hope it is enough for you to work off of as you write the code.

**Comment by sticnarf** [comment]
`kvs-client` and `kvs-server` are mistyped to `tikv-client` and `tikv-server`.

File: rust/projects/networking/project.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+  case the built-in engine is used, or "sled", in which case sled is used. If
+  this is the first run (there is no data previously persisted) then the default
+  value is "kvs"; if there is previously persisted data then the default is the
+  engine already in use. If data was previously persisted with a different
+  engine than selected, print an error and exit with a non-zero exit code.
+
+  Print an error and return a non-zero exit code on failure to bind a socket, if
+  `ENGINE-NAME` is invalid, if `IP-PORT` does not parse as an address.
+
+- `kvs-server -v`
```
Comment (sticnarf) [inline rust/projects/networking/project.md]:
```suggestion
- `kvs-server -V`
```

File: rust/projects/networking/project.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+  Print an error and return a non-zero exit code on server error,
+  or if `IP-PORT` does not parse as an address.
+
+- `kvs-client -V`
+
+  Print the version.
+
+The `kvs` library contains three types: `KvsClient`, `KvsServer`, and
+`KvsStore`. `KvsClient` implements the functionality required for `kvs-client`
```
Comment (Hoverbear) [inline rust/projects/networking/project.md]: This list might be better off as a true list.

File: rust/projects/networking/project.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+`TcpStream`. Before thinking about the protocol, modify `kvs-server` to listen
+for and accept connections, and `kvs-client` to initiate connections.
+
+For this project, the server is synchronous and single-threaded. That means that
+you will listen on a socket, then accept connections, and execute and respond to
+commands one at a time. In the future we will re-visit this decision multiple
+times on our journey toward an asynchronous, multi-threaded, and
+high-performance database.
+
+Thank about your manual testing workflow. Now that there are two executables to
```
Comment (Hoverbear) [inline rust/projects/networking/project.md]:
```suggestion
Think about your manual testing workflow. Now that there are two executables to
```

File: rust/projects/networking/project.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+deal with, you'll need a way to run them both at the same time. If you are like
+many, you will use two terminals, running `cargo run --bin --tikv-server` in
+one, where it runs until you press CTRL-D`, and `cargo run --bin --tikv-client`
```
Comment (Hoverbear) [inline rust/projects/networking/project.md]:
```suggestion
one, where it runs until you press CTRL-D, and `cargo run --bin --tikv-client`
```

File: rust/projects/networking/project.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+## Part 4: Implementing commands across the network
+
+Now it's time to implement the key/value store over the network, remotely
+executing commands that until now have been implemented within a single process.
+As with the file I/O you worked on in the last project to create the log, you
+will be serializing and streaming commands with the `Read` and `Write` traits.
+
+You are going to design a network protocol.
+
+TODO: Can't think of much to say about this
```
Comment (Hoverbear) [inline rust/projects/networking/project.md]: Maybe point them to the HTTP/1.1 examples on Wikipedia? Implementing super minimal HTTP/1.1 is very easy and I imagine many will do so anyways.

File: rust/projects/networking/project.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+
+## Part 5: More error handling
+
+- handle error responses by converting errors to a serializable format
```
Comment (Hoverbear) [inline rust/projects/networking/project.md]: This is missing the nice hand holding of the other parts. :) Maybe we could expand it a bit?

Comment (brson) [reply rust/projects/networking/project.md]: Oh, yeah. I just skipped it since I was more interested in the next. I'll put in a todo and fill it in later.

**Comment by brson** [comment]
Like project 2, this one is not complete, but I'm moving on to the next project. The rest will need to be done in future passes.

**Comment by brson** [comment]
Thanks for the reviews. All comments addressed.

**Pull request merged**